pkg/apis/core/validation/validation.go

```diff
@@ -6974,3 +7048,265 @@ func sameLoadBalancerClass(oldService, service *core.Service) bool {
+	allErrs = append(allErrs, ValidateResourceClaim(resourceClaim)...)
+	return allErrs
+}
+
+// ValidateResourceClaimStatusUpdate tests if an update to the status of a ResourceClaim is valid.
+func ValidateResourceClaimStatusUpdate(resourceClaim, oldResourceClaim *core.ResourceClaim) field.ErrorList {
+	allErrs := ValidateObjectMetaUpdate(&resourceClaim.ObjectMeta, &oldResourceClaim.ObjectMeta, field.NewPath("metadata"))
+	fldPath := field.NewPath("status")
+	// The name might not be set yet.
+	if resourceClaim.Status.DriverName != "" {
+		allErrs = append(allErrs, validateResourceDriverName(resourceClaim.Status.DriverName, fldPath.Child("driverName"))...)
+	} else if resourceClaim.Status.Allocation != nil {
+		allErrs = append(allErrs, field.Required(fldPath.Child("driverName"), "must be set when a claim is allocated"))
+	}
+
+	allErrs = append(allErrs, validateAllocationResult(resourceClaim.Status.Allocation, fldPath.Child("allocation"))...)
+	allErrs = append(allErrs, validateSliceIsASet(resourceClaim.Status.ReservedFor, v1.ResourceClaimReservedForMaxSize,
+		validateResourceClaimUserReference, fldPath.Child("reservedFor"))...)
+
+	// Now check for invariants that must be valid for a ResourceClaim.
+	if len(resourceClaim.Status.ReservedFor) > 0 {
+		if resourceClaim.Status.Allocation == nil {
+			allErrs = append(allErrs, field.Forbidden(fldPath.Child("reservedFor"), "only allocated claims can get reserved"))
+		} else {
+			if !resourceClaim.Status.Allocation.ShareableResource && len(resourceClaim.Status.ReservedFor) > 1 {
+				allErrs = append(allErrs, field.Forbidden(fldPath.Child("reservedFor"), "the claim cannot be reserved more than once"))
+			}
+			// Items may be removed from ReservedFor while the claim is meant to be deallocated,
+			// but not added.
+			if resourceClaim.DeletionTimestamp != nil ||
+				resourceClaim.Status.DeallocationRequested {
+				oldSet := sets.New(oldResourceClaim.Status.ReservedFor...)
+				newSet := sets.New(resourceClaim.Status.ReservedFor...)
+				newItems := newSet.Difference(oldSet)
+				if len(newItems) > 0 {
+					allErrs = append(allErrs, field.Forbidden(fldPath.Child("reservedFor"), "new entries may not get added while the claim is meant to be deallocated"))
+				}
+			}
+		}
+	}
+
+	if !oldResourceClaim.Status.DeallocationRequested &&
+		resourceClaim.Status.DeallocationRequested &&
+		len(resourceClaim.Status.ReservedFor) > 0 {
+		allErrs = append(allErrs, field.Forbidden(fldPath.Child("deallocationRequested"), "deallocation cannot be requested for claims which are in use"))
+	}
+
+	// Once deallocation has been requested, that request cannot be removed
+	// anymore because the deallocation may already have started. The field
+	// can only get reset by the driver together with removing the
+	// allocation.
+	if oldResourceClaim.Status.DeallocationRequested &&
+		!resourceClaim.Status.DeallocationRequested &&
+		resourceClaim.Status.Allocation != nil {
```

Comment:
Can I set allocation to nil without setting deallocationRequested to false? Is that covered?

Reply:
It makes sense to prevent that, will add it.

Reply:
Done.

I'm also preventing setting deallocationRequested when there is no allocation. That would have been handled by the driver by resetting it, but it makes more sense to not let a claim get into such a state in the first place - it's not needed.